experiment/build/generate_cloudbuild.py

```diff
@@ -13,28 +13,69 @@
 # limitations under the License.
 """Generates Cloud Build specification"""
 
-import argparse
-import json
-import yaml
+import os
 
 from experiment.build import docker_images
 
+BASE_TAG = 'gcr.io/fuzzbench'
```

Comment:
I don't like how this code uses a template to generate another template (a YAML file where variables are substituted).
If you use `tempfile` this code can generate exactly what gcb_build needs each time it needs to build something and it won't need to worry about substitutions. Not sure if we need to change this substitution stuff now in this PR but at least leave it as a TODO please.

Reply:
Noted. I can get rid of the fuzzer and benchmark substitutions, but not REPO and EXPERIMENT; its not contingent on using tempfile though.

Reply:
Sorry, why can't you get rid of them? Getting rid of half of them doesn't seem like progress since we are still generating a template.

Reply:
The `generate_cloudbuild.py` generates the spec first and is then passed to the gcloud command where the presence of REPO and EXPERIMENT is asserted. I didn't shift that assert to `generate_cloudbuild.py` because the function there only works on the given set of images.

Reply:
>where the presence of REPO and EXPERIMENT is asserted

Do you mean this: https://github.com/google/fuzzbench/pull/615/files#diff-3d114a89204a390da82e690ac453c840R101

> I didn't shift that assert to generate_cloudbuild.py because the function there only works on the given set of images.

Can you clarify what you mean here please?

I don't see why repo and experiment can't be grabbed in generate_cloudbuild.
The assertions are only relevant to substitutions which we won't be using anymore.

Reply:
Okay, I understand. I meant that I didn't grab it in generate_makefile since that is called in a loop, but I realize now that substitutions follow the same. I'll add that.

Reply:
Done

Reply:
This breaks tests of dispatcher, builder, etc which call gcb_build.py. Should I change experiment_utils.py to default to certain environment variables, or should I add a flag which sets the environment vars to test params and have the tests call that.

Reply:
>Should I change experiment_utils.py to default to certain environment variables

No.

>should I add a flag which sets the environment vars to test params and have the tests call that.

I don't know what this means. Where would you set this flag? What is "test params"? Are you saying if you pass a flag to the functions they call in gcb those get passed to these functions?

many tests deal with environment variables. They use the `environ` fixture and then set the vars they need.

Reply:
So, I grab the env vars in generate_cloudbuild.py which is imported by gcb_build.py. This causes the other tests to fail when they import gcb_build.py

To let the tests pass, is it fine if I make them return fuzzbench-test and test_experiment for CLOUD_PROJECT and EXPERIMENT in case of a KeyError?

Reply:
OK, this is a common problem.
The solution is to not make those values constants. You can create a helper function to get them from the env and use those instead of the constants.

Reply:
If you do this, then the usual solutions, like mock and environ work

Reply:
Fixed

Reply:
I replaced their use by calls. They were just 2 for each variable. Seemed easier than creating helper. Would you still have helper instead ?

Reply:
Could you link me to the line? I can't seem to find it. Thanks!